tests/EndToEndTests.hs

```diff
@@ -152,3 +155,137 @@ tests = testSpec "End-to-end tests" $ do
+      let expected =
+            object
+            [ "data" .= object
+              [ "dog" .= object
+                [ "name" .= ("Mortgage" :: Text)
+                ]
+              ]
+            ]
+      toJSON (toValue response) `shouldBe` expected
+    it "Allows calling query by name" $ do
+      let root = pure (viewServerDog mortgage)
+      let query = [r|query myQuery {
+                      dog {
+                        name
+                      }
+                    }
+                   |]
+      let Right name = makeName "myQuery"
+      response <- interpretQuery @QueryRoot root query (Just name) mempty
+      let expected =
+            object
+            [ "data" .= object
+              [ "dog" .= object
+                [ "name" .= ("Mortgage" :: Text)
+                ]
+              ]
+            ]
+      toJSON (toValue response) `shouldBe` expected
+    describe "Handles variables" $ do
+      let root = pure (viewServerDog mortgage)
+      let Right query =
+            compileQuery [r|query myQuery($whichCommand: DogCommand) {
+                              dog {
+                                name
+                                doesKnowCommand(dogCommand: $whichCommand)
+                              }
+                            }
+                           |]
+      it "Errors when no variables provided" $ do
+        response <- executeQuery  @QueryRoot root query Nothing mempty
+        let expected =
+              object
+              [ "data" .= object
+                [ "dog" .= object
+                  [ "name" .= ("Mortgage" :: Text)
+                  , "doesKnowCommand" .= Null
+                  ]
+                ]
+              , "errors" .=
+                [
+                  object
+                  -- TODO: This error message is pretty bad. We should define
+                  -- a typeclass for client-friendly "Show" (separate from
+                  -- actual Show which remains extremely useful for debugging)
+                  -- and use that when including values in error messages.
+                  [ "message" .= ("Could not coerce Name {unName = \"dogCommand\"} to valid value: ValueScalar' ConstNull not an enum: [Right (Name {unName = \"Sit\"}),Right (Name {unName = \"Down\"}),Right (Name {unName = \"Heel\"})]" :: Text)
+                  ]
+                ]
+              ]
+        toJSON (toValue response) `shouldBe` expected
+      it "Substitutes variables when they are provided" $ do
+        -- TODO: This is a crummy way to make a variable map. jml doesn't want
+        -- to come up with a new API in this PR, but probably we should have a
+        -- very simple function to turn a JSON value / object into the
+        -- variable map that we desire. Alternatively, we should have APIs
+        -- like Aeson does.
+        let Right varName = makeName "whichCommand"
+        let vars = Map.singleton (Variable varName) (toValue Sit)
```

Comment:
Gotcha, thanks for clarifying.

I guess this conversation has made it clear to me that `Enum "Foo"` is pretty much only an annotation for the schema plus a mean of getting all the possible values. I wonder whether our existing typeclass is doing too much: why should it care about conversion at all?